nvflare/apis/impl/client_controller.py

```diff
@@ -0,0 +1,223 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, Union
+
+from nvflare.apis.client import Client
+from nvflare.apis.controller_spec import ClientTask, ControllerSpec, SendOrder, Task, TaskCompletionStatus
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_constant import FLContextKey, ReservedKey, ReservedTopic, ReturnCode, SiteType
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.security.logging import secure_format_exception
+
+
+class ClientController(FLComponent, ControllerSpec):
```

Comment:
How are we going to handle the failure?
Will we have something like "snapshot" on the client side?

Reply:
no, the snapshot is used for the job HA execution. client controller is performed in the client executor.

Reply:
so if client use swarm learning and 8 out of 10 clients finish the training, if it does not complete one round, something wrong happens we might lose all training progress.

We could mark this as future enhancements.

Reply:
All the client results are included in the "replies", including the failed client. The calling logic can make use of the replies to decide the handling logic of the failed clients.